### DISCUSSION TITLE
How to calculate CPU usage per method from JFR file

### DISCUSSION THREAD

**Discussion by 1v0dev** [async-profiler/async-profiler#520] [Q&A] [answered]
Hi,
I am profiling an app with both CPU and alloc. The result is an JFR file which contains events with stack traces in them. The events have start and end time, but in my file, they are exactly the same for all ```jdk.ExecutionSample``` events. Can someone explain how I can extract or calculate the CPU usage or at least time elapsed per event/stack trace/method from the JFR file so I can put it into a database? I am parsing the JFR inside Java program with the standard JDK api.

**Comment by toktarev** [comment]
Hello.

You could probably convert the JFR file to a flame graph (https://github.com/jvm-profiling-tools/async-profiler#flame-graph-visualization)?


Converters between profile formats: converter.jar
(JFR to Flame Graph, JFR to FlameScope, collapsed stacks to Flame Graph)

**Reply by 1v0dev** [reply]
Thank you. But I want to have the raw data so I can insert it into a database and run my own analysis on it. Converting to FlameGraph and then trying to parse that graph looks like workaround to me. Since I can directly open JFR in Java and obviously the data exist in the JFR, I should be able to parse it directly instead of first converting to something else. Reverse engineering the coverter.jar to see how it does the calculations is an option, but I hope someone know how it's done and save me the time.

**Comment by krzysztofslusarski** [comment]
The JFR format contains stacktraces dumped with sampling. A single event of a ```jdk.ExecutionSample``` type means that at the ```startTime``` there was a thread with such a stacktrace. There is no information about duration in a single ```jdk.ExecutionSample```. 

If you want a CPU usage it is covered by ```jdk.CPULoad``` events. You need async-profiler 2.* to gather them.

**Reply by 1v0dev** [reply]
Thank you for the explanation. I opened the JFR with Java Mission Control and was under the impression that they show processing time and calculate it somehow in relation of the CPU Load events. But in reality, it is using the number of executions to assign percentage for each method.

**Comment by apangin** [comment, marked as answer]
To begin with, async-profiler is a *sampling* profiler. It does not measure execution time (CPU usage, whatever) of individual methods. Instead, it periodically takes a stack trace of the code running *right now*. In that way, each sample is a momentary snapshot, it has no length or duration.

However, basing on the number of samples, we may reason about the "weight" of each method (or a stack trace) in the profile. For example, if the sampling period is 10ms of CPU time, and some method occurs 500 times among all samples, we may *estimate* the CPU time of this method as 500*10ms = 5 seconds.

Now about the formats. JFR format is specifically for recording all *individual* samples. It does not aggregate samples by design - it's the job of the visualization tools. JMC and the built-in Flame Graph converter (converter.jar) are the examples of such tools.

If you are interested in per-method aggregated statistics, you may try different output formats, e.g. `collapsed`, which is basically the plain text format, very simple for parsing and processing.

If you still want to use JFR and aggregate the required statistics manually, look at [JfrReader](https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/converter/one/jfr/JfrReader.java) in async-profiler sources. [jfr2flame](https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/converter/jfr2flame.java) is a working example on how to use `JfrReader` and `EventAggregator` for JFR processing.

**Reply by 1v0dev** [reply]
Thank you very much! Especially for the logic to estimate the CPU time. I already started looking through ```JFRReader``` and seems it will help me a lot.
